src/sage/rings/polynomial/multi_polynomial_ring_base.pyx

```diff
@@ -1794,3 +1830,16 @@ def unpickle_MPolynomialRing_generic(base_ring, n, names, order):
     from .polynomial_ring_constructor import PolynomialRing
 
     return PolynomialRing(base_ring, n, names=names, order=order)
+
+class MatrixPolynomialAction(Action):
```

Comment:
Each of the methods for this class needs at least one doctest. For the `__init__`, I recommend creating an instance `A` and then simply running `TestSuite(A).run()`.

Reply:
Hi, I've tried this but I'm getting the following error

Failure in _test_pickling:
Traceback (most recent call last):
  File "/home/bruno/Documents/sage/src/sage/misc/sage_unittest.py", line 297, in run
    test_method(tester=tester)
  File "sage/structure/sage_object.pyx", line 652, in sage.structure.sage_object.SageObject._test_pickling
    tester.assertEqual(loads(dumps(self)), self)
  File "sage/misc/persist.pyx", line 995, in sage.misc.persist.loads
    ans = unpickler.load()
  File "sage/rings/polynomial/multi_polynomial_ring_base.pyx", line 1836, in sage.rings.polynomial.multi_polynomial_ring_base.MatrixPolynomialAction.__init__
    def __init__(self, MS, PR):
TypeError: __init__() takes exactly 3 positional arguments (5 given)
------------------------------------------------------------
The following tests failed: _test_pickling


any pointers?

Reply:
It was a good thing to test this then. You can do one of two things:

1. Implement is a `__reduce__` method; just copy the one from `Action` with removing the unnecessary inputs.
2. Add `is_left` and `op` as inputs that are ignored to your `__init__`.

I would strongly recommend the first one; it is a simple one-line function (with a simple test).

Reply:
Hi I've just tried this but it's throwing quite a strange error now:

```
File "src/sage/rings/polynomial/multi_polynomial_ring_base.pyx", line 1848, in sage.rings.polynomial.multi_polynomial_ring_base.MatrixPolynomialAction.__init__
Failed example:
    TestSuite(A).run()
Expected nothing
Got:
    Failure in _test_pickling:
    Traceback (most recent call last):
      File "/home/bruno/Documents/sage/src/sage/misc/sage_unittest.py", line 297, in run
        test_method(tester=tester)
      File "sage/structure/sage_object.pyx", line 652, in sage.structure.sage_object.SageObject._test_pickling
        tester.assertEqual(loads(dumps(self)), self)
      File "/home/bruno/Documents/sage/local/var/lib/sage/venv-python3.11.1/lib/python3.11/unittest/case.py", line 873, in assertEqual
        assertion_func(first, second, msg=msg)
      File "/home/bruno/Documents/sage/local/var/lib/sage/venv-python3.11.1/lib/python3.11/unittest/case.py", line 866, in _baseAssertEqual
        raise self.failureException(msg)
    AssertionError: Left [77 chars]riate Polynomial Ring in w, x, y, z over Finite Field of size 2 != Left [77 chars]riate Polynomial Ring in w, x, y, z over Finite Field of size 2
    ------------------------------------------------------------
    The following tests failed: _test_pickling
```
I'll commit the change though and you can have a look

Reply:
Thank you. This is something slightly annoying I didn't think about. You need to also implement a comparison method:
```python
def __eq__(self, other):
    return (type(self) == type(other)
                  and self.MS == (<MatrixPolynomialAction> other).MS
                  and self.PR == (<MatrixPolynomialAction> other).PR)
```
which also means implementing a hash
```python
def __hash__(self):
    return hash((self,MS, self.PR))
```
Nothing will actually be comparing these actions, but hashing and pickling are very commonly invoked. Anyways, this should now make all the tests pass.

Reply:
Hi I've got it working but your `__eq__` implementation threw this error

```
[sagelib-10.2.beta0]     Error compiling Cython file:
[sagelib-10.2.beta0]     ------------------------------------------------------------
[sagelib-10.2.beta0]     ...
[sagelib-10.2.beta0]             substitution_dict = {v: s for v, s in zip(self._poly_vars, vars_to_sub_ring_context)}
[sagelib-10.2.beta0]             return polynomial.subs(substitution_dict)
[sagelib-10.2.beta0] 
[sagelib-10.2.beta0]         def __eq__(self, other):
[sagelib-10.2.beta0]             return (type(self) == type(other)
[sagelib-10.2.beta0]                     and self.MS == (<MatrixPolynomialAction> other).MS
[sagelib-10.2.beta0]                                     ^
[sagelib-10.2.beta0]     ------------------------------------------------------------
[sagelib-10.2.beta0] 
[sagelib-10.2.beta0]     sage/rings/polynomial/multi_polynomial_ring_base.pyx:1865:33: 'MatrixPolynomialAction' is not a type identifier
```

so I replaced it with 

```    
def __eq__(self, other):
        if isinstance(other, MatrixPolynomialAction):
            return self.MS == other.MS and self.PR == other.PR
        return False
```

hopefully this is roughly the same

Reply:
Ah, sorry, it is not a `cdef` class like I was thinking. Your implementation is fine.